### PR TITLE
[CALCITE-6649] Enhance RelMdPredicates pull up predicate from PROJECT

### PR DESCRIPTION
**Pull request by NobiGo**

### CODE REVIEW

**Review by mihaibudiu — APPROVED** [review]
I hope this is not too expensive; some of these objects are reallocated potentially many times.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java [hunk 1]
```diff
@@ -200,15 +204,17 @@ public RelOptPredicateList getPredicates(Project project,
             project.getRowType().getFieldCount());
 
     for (Ord<RexNode> expr : Ord.zip(project.getProjects())) {
-      if (expr.e instanceof RexInputRef) {
-        int sIdx = ((RexInputRef) expr.e).getIndex();
-        m.set(sIdx, expr.i);
+      RexNode rexNode = simplify.simplify(expr.e);
+      Ord<RexNode>  newExpr = new Ord<>(expr.i, rexNode);
+      if (newExpr.e instanceof RexInputRef) {
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java:209]: Maybe adding some javadoc would be better

**Review by caicancai — APPROVED** [review]
LGTM

**Comment by NobiGo** [comment]
This PR has been fixed by CALCITE-6586. I only added some tests to confirm we don't need to optimize this in RelMdPredicates.

**Comment by NobiGo** [comment]
I will merge this PR in the next 48 hours if there are no more comments.

**Pull request merged**